web-api/src/personalization/routes.rs

```diff
@@ -414,6 +420,59 @@ pub(crate) async fn personalize_documents_by(
     Ok(Some(all_documents))
 }
 
+#[derive(Deserialize)]
+struct SemanticSearchQuery {
+    count: Option<usize>,
+}
+
+impl SemanticSearchQuery {
+    fn document_count(&self, config: &SemanticSearchConfig) -> Result<usize, Error> {
+        let count = self.count.map_or(config.default_number_documents, |count| {
+            count.min(config.max_number_documents)
+        });
+
+        if count > 0 {
+            Ok(count)
+        } else {
+            Err(BadRequest::from("count has to be at least 1").into())
+        }
+    }
+}
+
+#[derive(Serialize)]
+struct SemanticSearchResponse {
+    documents: Vec<PersonalizedDocumentData>,
+}
+
+async fn semantic_search(
+    state: Data<AppState>,
+    document_id: Path<String>,
+    query: Query<SemanticSearchQuery>,
+) -> Result<impl Responder, Error> {
+    let document_id = document_id.into_inner().try_into()?;
+    let count = query.document_count(state.config.as_ref())?;
+
+    let embedding = storage::Document::get_embedding(&state.storage, &document_id)
+        .await?
+        .ok_or(DocumentNotFound)?;
+
+    let documents = storage::Document::get_by_embedding(
+        &state.storage,
+        KnnSearchParams {
+            excluded: &[document_id],
+            embedding: &embedding,
+            k_neighbors: count,
+            num_candidates: count,
+            published_after: None,
+        },
+    )
+    .await?;
+
+    Ok(Json(SemanticSearchResponse {
+        documents: documents.into_iter().map(Into::into).collect(),
```

Comment:
🎨 `Itertools` provides a shorthand for that
```suggestion
        documents: documents.into_iter().map_into().collect(),
```